x/ibc/core/04-channel/types/packet.go

```diff
@@ -1,28 +1,42 @@
 package types
 
 import (
-	"github.com/tendermint/tendermint/crypto/tmhash"
+	"crypto/sha256"
 
+	"github.com/cosmos/cosmos-sdk/codec"
 	sdk "github.com/cosmos/cosmos-sdk/types"
 	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
 	clienttypes "github.com/cosmos/cosmos-sdk/x/ibc/core/02-client/types"
 	host "github.com/cosmos/cosmos-sdk/x/ibc/core/24-host"
 	"github.com/cosmos/cosmos-sdk/x/ibc/core/exported"
 )
 
-// CommitPacket returns a packet commitment bytes. The commitment consists of:
-// hash(timeout_timestamp + timeout_version_number + timeout_version_height + data) from a given packet.
-func CommitPacket(packet exported.PacketI) []byte {
+// CommitPacket returns the packet commitment bytes. The commitment consists of:
+// sha256_hash(timeout_timestamp + timeout_height.VersionNumber + timeout_height.VersionHeight + sha256_hash(data))
```

Comment:
cc @cwgoes

Reply:
> Isn't it still possible for distinct packets to have the same commitment hash? Eg. if the same sender sends two versions of the same FungibleTokenTransfer with the same timeout? The packets themselves will differ by the sequence number, but the CommitPacket will be the same since it only includes the timeout and packet data. Is this intended?

Yes, it is still possible for two distinct packets to have the same commitment hash, but only if they have the same data (modulo being able to find a collision) - it's two packets with different data but the same hash that we would be concerned about. 

> It seems this might be ok because these CommitPacket hashes are stored under keys that include the source channel and sequence number (and everything is ultimately verified by a Client), but it's still a bit surprising.

Yes, I don't think this is a security issue. I agree that it's potentially surprising but this hash is really only intended to be provide verification of the packet _data_ - semantically, we could store the data (and timestamp) directly, but if the packet data is large that might be expensive so we just store the hash instead.

> I guess the reason not to just hash the whole packet in the first place is because of possible protobuf shenanigans? I'm not sure the state of a more general strategy in the SDK for dealing with that. But in the meantime should more fields from the packet be reflected in the pre-image? Say, at least the sequence number? Or is there intended benefit from these spurious CommitPacket collisions?

Hmm. See response above - if this is intended only to be a way to verify the packet _data_, which was the original intent, it would be confusing to me to hash additional fields. I suppose hashing additional fields certainly isn't a bad idea, but it makes things a bit less clear in my opinion because other fields (e.g. sequence numbers) are being verified separately (and must be verified separately, so there can be only one packet per sequence number, for example).

Do you think clearer documentation here on the purpose of this hash and what it needs to verify would be enough? Or do you think we should reconsider how we're authenticating all of the packet fields (possible, but probably for a future version)?